source/slang/slang-ir-link.cpp

```diff
@@ -1606,6 +1606,40 @@ LinkedIR linkIR(
 
     context->builder->setInsertInto(context->getModule()->getModuleInst());
 
+    // Raypayload/related objects found under GLSL are likley never referenced
+    // this is because internally they are used by slang functions; these emit later in code gen
+    // as a result we emit them now
+    // we also emit these early since if these are not at the top of scope, that could be deadly to a compile
+    if (targetProgram->getOptionSet().getBoolOption(CompilerOptionName::AllowGLSL)) 
```

Comment:
Instead of doing this in linking, can we just mark the global variables of those payload as [KeepAlive] and let linking copy them in?

Reply:
I tested and looked around the code: 
KeepAliveDecoration allows the var to survive optimizations, but the initial linking will still never copy the raytracing object in GLSL syntax since the var will likley never be referenced.

we don't need the `addInst` though since once cloned the inst will be in the new context and will add later anyways; I think we should still add to keep the code working more like `insertGlobalValueSymbols`

Reply:
You also need a `kIROp_HLSLExportDecoration` on the varible to make it survive linking regardless of whether it has uses.

Reply:
See slang-ir-link.cpp:1673 for reference.

Reply:
I don't need to clone for the attribute to live, but I realized the emitted object will be at the bottom of the program if not emitted early

Reply:
I checked, and this would be a fatal program error

Reply:
we need to early emit the raytracing objects to ensure they are at top of scope above everything else (including stdlib functions)

Reply:
The IR is designed to allow arbitrary odering at the global scope level, so I wonder why the implementation requires those objects to be on top, which seems worrying.

Reply:
raytracing intrinsic's resolve at just before emitting stage (used to be during emitting stage) 

due to this since we don't reference raytracing objects normally in GLSL, they are placed at the bottom 

we can change this behavior reliably by making all raytracing objects 'dependent' on all raytracing intrinsic I suppose

Reply:
This means that we need to either
1) update the emit logic in emit-glsl to always emit payload variables before anything else;  or 
2) add some decorations to establish that dependency, such as decorate the call to the `traceRay` function with a reference to the payload variable.

We should not write code based on assumption that things in global scope is in any particular order.

Reply:
Btw, why is it a problem if it is declared at the bottom? Does GLSL not work in that case?

Reply:
that's what the code currently does, it ensures the ray objects are emitted directly after all hashed strings, at top of scope

Reply:
> Btw, why is it a problem if it is declared at the bottom? Does GLSL not work in that case?

likely does not work: https://shader-playground.timjones.io/788a40ca09027e0f1be0fddefe447dee

Reply:
We can't assume that the order of global insts are preserved across passes, so just placing them at the beginning of the module at linking does not guarantee they will continue to be at top when we get to emit.
That's why I am saying we should not make assumptions on the ordering of global insts.

If you want to implement 1), you need to do so in emit-glsl.cpp to emit those global variables first, then emit entrypoint.

Reply:
You example is different from what we are talking about here. The issue we are discussing is that there is no direct reference to that global variable in the glsl, and it is accessed indirectly through an index in code. So the GLSL should be if the variable itself is defined last.

Reply:
I am changing the example to reflect as such

Reply:
does not work, order matters

Reply:
> We can't assume that the order of global insts are preserved across passes, so just placing them at the beginning of the module at linking does not guarantee they will continue to be at top when we get to emit. That's why I am saying we should not make assumptions on the ordering of global insts.
> 
> If you want to implement 1), you need to do so in emit-glsl.cpp to emit those global variables first, then emit entrypoint.

I will do this (this is what the code originally did, so I can likley just use that code)

Reply:
OK, then we need to modify `slang-emit-c-like.cpp:: CLIkeSourceEmitter::computeEmitActions()` to visit all ray tracing globals first.

Reply:
we would also need to modify the SPIR-V emitter(?)

Reply:
I don't think the SPIRV needs to change. Should be fine to have out of order spirv. Will need to see if that passes validation.

Reply:
But in case SPIRV does need it, the logic is the same there. You can add calls to `ensureInst` on those globals in `emitSPIRVFromIR()`.

Reply:
We already have some logic that to ensure certain types of insts are emitted before the others.

Reply:
I will implement this shortly 👍, I will just adjust slang-ir-glsl-legalize first so we can use HLSL+GLSL syntax for raytracing before hand (I think it would be quite nice to have HLSL syntax access for if adding to legacy GLSL raytracing code)

Reply:
Continue this thread, if we make sure to also emit `HLSLExport` decoration when we lower IR for the global ray tracing object, it should just be cloned during linking without the explicit loop here. If this is not the case we need to figure out why.